akka-actor/src/main/scala/akka/actor/ActorSystem.scala

```diff
@@ -930,18 +951,48 @@ private[akka] class ActorSystemImpl(
   def registerOnTermination(code: Runnable): Unit = { terminationCallbacks.add(code) }
 
   override def terminate(): Future[Terminated] = {
+    if (terminating.compareAndSet(false, true)) {
+      val coordinatedShutdown = CoordinatedShutdown(this)
+      if (settings.CoordinatedShutdownRunByActorSystemTerminate &&
+          coordinatedShutdown.shutdownReason().isEmpty && !aborting) {
+
+        if (!settings.CoordinatedShutdownTerminateActorSystem) {
+          // Configured to not terminate ActorSystem from CoordinatedShutdown, but now it's an explicit call
+          // to ActorSystem.terminate so adding that task so that the ActorSystem.finalTerminate is called at the end of
+          // the CoordinatedShutdown phases.
+          coordinatedShutdown.addTask(PhaseActorSystemTerminate, "terminate-actor-system") { () =>
+            finalTerminate()
+            whenTerminated.map(_ => Done)(ExecutionContexts.sameThreadExecutionContext)
+          }
+        }
+
+        if (!settings.CoordinatedShutdownTerminateActorSystem && coordinatedShutdown.shutdownReason().nonEmpty) {
+          // another thread has started CoordinatedShutdown and then it's not certain that the task added above
+          // made it and will be run, only safe option is to call finalTerminate() right away
+          finalTerminate()
+        } else
+          coordinatedShutdown.run(CoordinatedShutdown.ActorSystemTerminateReason)
+
```

Comment:
@raboof the problems are around when `actor-system-terminate-enabled = off` (settings.CoordinatedShutdownTerminateActorSystem) because then Coordinated shutdown will not call `finalTermiante`. That's why I try to add the task above, but I realized that it wasn't completely safe, as mentioned in comments above.

Note that there are two entry points `ActorSystem.terminate()` and `CoordinatedShutdown.run()`.

Reply:
I added the `terminating` AtomicBoolean so that it's protecting against to early `finalTermiante` for multiple calls to terminate(). Now that can only happen if ActorSystem.terminate() and CoordinatedShutdown.run() are called at the same time and `actor-system-terminate-enabled = off` is configured. That seems odd enough to be ok.

Reply:
Hmm, good observation.

It makes me wonder if we should allow the combination of `CoordinatedShutdownRunByActorSystemTerminate=true` and `CoordinatedShutdownTerminateActorSystem=false` - does this ever make sense?

Reply:
good thinking, I actually started by piggy backing on `CoordinatedShutdownTerminateActorSystem`, but was then thinking that it was separate things, mostly in case someone wanted the old behavior.

So, we can still have two properties, but disallow that combination. That will make this much simpler.

Reply:
disallowed the problematic combination: ac102ef